File: connectors/claude/claude_client.py
"""
OrgSuite Claude (Anthropic) client skeleton
Status: Ready to Configure / Requires Authorization

- Reads ONLY from os.environ["ANTHROPIC_API_KEY"]
- Never hard-codes secrets
- Official anthropic SDK only
- Error handling + structured logging
- Model pinned

Install: pip install anthropic
Env: ANTHROPIC_API_KEY=sk-ant-... (Vercel / Firebase / host only)

Linked to Linear: PSE-31
"""

from __future__ import annotations

import logging
import os
from typing import Any, List, Optional

from anthropic import Anthropic, APIError, APIStatusError, RateLimitError

# Pin model. Update only after checking Anthropic docs.
DEFAULT_MODEL = "claude-opus-4-6"  # example; replace with exact current ID
DEFAULT_MAX_TOKENS = 1024

logger = logging.getLogger("orgsuite.claude_client")
logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(name)s %(message)s")


class ClaudeClient:
    def __init__(self) -> None:
        api_key = os.environ.get("ANTHROPIC_API_KEY", "")
        self._api_key_present = bool(api_key and api_key.startswith("sk-ant-"))
        self._client: Optional[Anthropic] = None

        if not self._api_key_present:
            logger.warning("ANTHROPIC_API_KEY missing or invalid format. Client will fail on first call.")
            return

        # SDK reads ANTHROPIC_API_KEY from env by default when api_key is omitted.
        # We pass it only to make the source explicit.
        self._client = Anthropic(api_key=api_key)

    def is_configured(self) -> bool:
        """Health check: returns True only if key is present (no network call)."""
        return self._api_key_present

    def messages(
        self,
        messages: List[dict[str, str]],
        *,
        model: str = DEFAULT_MODEL,
        max_tokens: int = DEFAULT_MAX_TOKENS,
        system: Optional[str] = None,
        temperature: Optional[float] = None,
        request_id: Optional[str] = None,
    ) -> Any:
        """
        Send a messages request to Claude.
        Raises on missing key, validation, network, rate-limit, or API errors.
        """
        if not self._client or not self._api_key_present:
            msg = "ClaudeClient: ANTHROPIC_API_KEY is not set or invalid. Configure in environment only."
            logger.error("%s request_id=%s", msg, request_id)
            raise RuntimeError(msg)

        if not messages:
            raise ValueError("ClaudeClient: messages list is required and must not be empty")

        for m in messages:
            if not isinstance(m, dict) or "role" not in m or "content" not in m:
                raise ValueError("ClaudeClient: each message must be a dict with role and content")
            if not isinstance(m["content"], str) or len(m["content"]) > 200_000:
                raise ValueError("ClaudeClient: content must be a string under safe length")

        rid = request_id or f"claude-{id(self)}"
        logger.info(
            "Claude Messages API call starting request_id=%s model=%s max_tokens=%s count=%d",
            rid,
            model,
            max_tokens,
            len(messages),
        )

        try:
            response = self._client.messages.create(
                model=model,
                max_tokens=max_tokens,
                messages=messages,  # type: ignore[arg-type]
                system=system,
                temperature=temperature,
            )
            logger.info(
                "Claude Messages API call succeeded request_id=%s model=%s stop=%s in=%s out=%s",
                rid,
                getattr(response, "model", None),
                getattr(response, "stop_reason", None),
                getattr(getattr(response, "usage", None), "input_tokens", None),
                getattr(getattr(response, "usage", None), "output_tokens", None),
            )
            return response
        except RateLimitError as e:
            logger.warning("Claude rate limited request_id=%s – apply backoff: %s", rid, e)
            raise
        except APIStatusError as e:
            status = getattr(e, "status_code", None)
            if status in (401, 403):
                logger.error("Claude auth failure request_id=%s status=%s", rid, status)
            else:
                logger.error("Claude API status error request_id=%s status=%s: %s", rid, status, e)
            raise
        except APIError as e:
            logger.error("Claude API error request_id=%s: %s", rid, e)
            raise


# Module-level instance for simple import
claude_client = ClaudeClient()


if __name__ == "__main__":
    # Local smoke test only – never commit real key
    if claude_client.is_configured():
        print("Configured: yes (key present)")
    else:
        print("Configured: no – set ANTHROPIC_API_KEY in environment")

File: connectors/claude/claudeClient.ts
/**
 * OrgSuite Claude (Anthropic) client skeleton
 * Status: Ready to Configure / Requires Authorization
 *
 * - Reads ONLY from process.env.ANTHROPIC_API_KEY
 * - Never hard-codes secrets
 * - Official @anthropic-ai/sdk only
 * - Error handling + structured logging
 * - Model pinned; update when Anthropic deprecates
 *
 * Install: npm install @anthropic-ai/sdk
 * Env: ANTHROPIC_API_KEY=sk-ant-... (Vercel / Firebase / host only)
 *
 * Linked to Linear: PSE-31
 */

import Anthropic from "@anthropic-ai/sdk";
import { createLogger } from "./logger"; // replace with your OrgSuite logger

const log = createLogger("claude-client");

// Pin model. Update only after checking Anthropic docs.
const DEFAULT_MODEL = "claude-opus-4-6"; // example current Opus-class; replace with exact ID from console
const DEFAULT_MAX_TOKENS = 1024;

export interface ClaudeMessageInput {
  role: "user" | "assistant";
  content: string;
}

export interface ClaudeCallOptions {
  model?: string;
  maxTokens?: number;
  system?: string;
  temperature?: number;
  /** Optional OrgSuite correlation id for audit */
  requestId?: string;
}

export class ClaudeClient {
  private client: Anthropic | null = null;
  private readonly apiKeyPresent: boolean;

  constructor() {
    const apiKey = process.env.ANTHROPIC_API_KEY;
    this.apiKeyPresent = Boolean(apiKey && apiKey.startsWith("sk-ant-"));

    if (!this.apiKeyPresent) {
      log.warn("ANTHROPIC_API_KEY missing or invalid format. Client will fail on first call.");
      return;
    }

    // SDK automatically reads ANTHROPIC_API_KEY from env when not passed.
    // We pass it explicitly only to make the dependency obvious and avoid accidental browser use.
    this.client = new Anthropic({
      apiKey, // still from env only
      // dangerouslyAllowBrowser: false  // default; never enable
    });
  }

  /**
   * Send a messages request to Claude.
   * Throws on missing key, network, rate-limit, or API errors.
   */
  async messages(
    messages: ClaudeMessageInput[],
    options: ClaudeCallOptions = {}
  ): Promise<Anthropic.Messages.Message> {
    if (!this.client || !this.apiKeyPresent) {
      const err = new Error(
        "ClaudeClient: ANTHROPIC_API_KEY is not set or invalid. Configure in environment only."
      );
      log.error({ err, requestId: options.requestId }, "Missing API key");
      throw err;
    }

    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error("ClaudeClient: messages array is required and must not be empty");
    }

    // Basic input sanitization (extend with your OrgSuite validator)
    for (const m of messages) {
      if (!m.role || !m.content || typeof m.content !== "string") {
        throw new Error("ClaudeClient: each message must have role and string content");
      }
      if (m.content.length > 200_000) {
        // rough guard; Anthropic has its own limits
        throw new Error("ClaudeClient: message content exceeds safe length");
      }
    }

    const model = options.model ?? DEFAULT_MODEL;
    const maxTokens = options.maxTokens ?? DEFAULT_MAX_TOKENS;
    const requestId = options.requestId ?? `claude-${Date.now()}`;

    log.info(
      { requestId, model, maxTokens, messageCount: messages.length },
      "Claude Messages API call starting"
    );

    try {
      const response = await this.client.messages.create({
        model,
        max_tokens: maxTokens,
        messages: messages.map((m) => ({
          role: m.role,
          content: m.content,
        })),
        system: options.system,
        temperature: options.temperature,
      });

      log.info(
        {
          requestId,
          model: response.model,
          stopReason: response.stop_reason,
          inputTokens: response.usage?.input_tokens,
          outputTokens: response.usage?.output_tokens,
        },
        "Claude Messages API call succeeded"
      );

      return response;
    } catch (error: unknown) {
      // Structured error handling
      const err = error as any;
      const status = err?.status ?? err?.statusCode;
      const type = err?.error?.type ?? err?.type;

      if (status === 429) {
        log.warn({ requestId, status, type }, "Claude rate limited – apply backoff");
        // Caller should implement exponential backoff / queue
      } else if (status === 401 || status === 403) {
        log.error({ requestId, status, type }, "Claude auth failure – check key / scopes");
      } else {
        log.error({ requestId, status, type, message: err?.message }, "Claude API error");
      }

      throw error; // re-throw for upstream handling
    }
  }

  /** Health check: returns true only if key is present (does not call network). */
  isConfigured(): boolean {
    return this.apiKeyPresent;
  }
}

// Singleton for simple use (or inject in your DI)
export const claudeClient = new ClaudeClient();

/**
 * Example usage (never run in production without key):
 *
 * const result = await claudeClient.messages(
 *   [{ role: "user", content: "Summarize OrgSuite connector status." }],
 *   { requestId: "pse-31-test", system: "You are a secure OrgSuite assistant." }
 * );
 * console.log(result.content);
 */
